servicetalk-opentelemetry-http/src/main/java/io/servicetalk/opentelemetry/http/GrpcSemanticAttributesExtractor.java

```diff
@@ -0,0 +1,115 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.servicetalk.opentelemetry.http;
+
+import io.servicetalk.transport.api.ConnectionInfo;
+
+import io.opentelemetry.api.common.AttributeKey;
+import io.opentelemetry.api.common.AttributesBuilder;
+import io.opentelemetry.context.Context;
+import io.opentelemetry.instrumentation.api.instrumenter.AttributesExtractor;
+
+import java.net.InetSocketAddress;
+import java.net.SocketAddress;
+import javax.annotation.Nullable;
+
+/**
+ * Extractor for gRPC semantic convention attributes using stable OpenTelemetry APIs.
+ * <p>
+ * This class manually implements RPC semantic conventions as defined by OpenTelemetry,
+ * avoiding dependencies on alpha/incubator APIs.
+ */
+final class GrpcSemanticAttributesExtractor implements AttributesExtractor<RequestInfo, GrpcTelemetryStatus> {
+
+    // RPC semantic convention attribute keys
+    private static final AttributeKey<String> RPC_SYSTEM = AttributeKey.stringKey("rpc.system");
+    private static final AttributeKey<String> RPC_SERVICE = AttributeKey.stringKey("rpc.service");
+    private static final AttributeKey<String> RPC_METHOD = AttributeKey.stringKey("rpc.method");
+    private static final AttributeKey<Long> RPC_GRPC_STATUS_CODE = AttributeKey.longKey("rpc.grpc.status_code");
+
+    static final AttributeKey<String> NETWORK_PEER_ADDRESS = AttributeKey.stringKey("network.peer.address");
+    static final AttributeKey<Long> NETWORK_PEER_PORT = AttributeKey.longKey("network.peer.port");
+
+    static final GrpcSemanticAttributesExtractor INSTANCE = new GrpcSemanticAttributesExtractor();
+
+    private GrpcSemanticAttributesExtractor() {
+    }
+
+    @Override
+    public void onStart(AttributesBuilder attributes, Context parentContext, RequestInfo request) {
+        // Set the RPC system to gRPC
+        attributes.put(RPC_SYSTEM, "grpc");
+
+        // Extract service and method from gRPC path: /service.name/MethodName
+        String path = request.getMetadata().path();
+        extractServiceAndMethod(path, attributes);
+        extractPeerAddress(request, attributes);
+    }
+
+    @Override
+    public void onEnd(AttributesBuilder attributes, Context context, RequestInfo request,
+                      @Nullable GrpcTelemetryStatus response, @Nullable Throwable error) {
+        // Extract gRPC status code from response headers/trailers (trailers first per gRPC spec)
+        if (response != null) {
+            extractGrpcStatusCode(response, attributes);
+        }
+    }
+
+    private static void extractServiceAndMethod(String path, AttributesBuilder attributes) {
+        if (path == null || path.isEmpty()) {
+            return;
+        }
+
+        // gRPC path format: /service.name/MethodName
+        // Remove leading slash if present
+        String normalizedPath = path.startsWith("/") ? path.substring(1) : path;
```

Comment:
The first character can be checked with `path.charAt(0) == '/'`. Also, consider using indexes to avoid an extra `substring` here, you can just do `substring(1, lastSlashIndex)` later when you extract `service`